buzzmobile/tests/rostest_utils.py

```diff
@@ -0,0 +1,161 @@
+        self.roscore = subprocess.Popen(['roscore'])
+        old_setup(self)
+
+    def new_teardown(self): 
+        old_teardown(self)
+        self.roscore.kill()
+        subprocess.call(['killall', '-9', 'rosmaster'])
+        self.roscore.wait()
+
+    obj.setUp = new_setup
+    obj.tearDown = new_teardown
+    return obj
+
+
+class ROSLauncher(roslaunch.scriptapi.ROSLaunch):
+    """
+    ROSLaunch but allows the use of launch files.
+
+    This was found by peering into the roslaunch util source code from
+    `which roslaunch`. 
+    """
+    def __init__(self, files):
+        super().__init__()
+        uuid = roslaunch.rlutil.get_or_generate_uuid(None, True) 
+        self.parent = roslaunch.parent.ROSLaunchParent(uuid, files, is_core=False) 
+
+# TODO(joshuamorton@gatech.edu): swap this to a Map[int, launcher], where the
+# int is incrementally generated and stored within the calling decorator. This
+# will prevent the issue of overwriting or cross writing if it becomes a
+# danger.
+_launcher = None
+
+def with_launch_file(package, launch):
+    """Decorator to source a launch file for running nodes.
+
+    This should always be run first.
+
+    This and launch nodes work together gracefully, but this poses a danger.
+    Because they rely on a global variable, multiple running simultaneously
+    (in the same thread) can cause issues by overwriting the 'launcher'
+    value. This could be fixed if needed, but I don't think it will be an
+    issue.
+    """
+    f = roslaunch.rlutil.resolve_launch_arguments([package, launch])
+    def launcher(func):
+        @functools.wraps(func)
+        def new_test(self):
+            launch = ROSLauncher(f)
+            launch.start()
+            global _launcher
+            _launcher = launch
+
+            temp = func(self)
+            _launcher = None
+            return temp
+        return new_test
+    return launcher
+            
+def launch_node(package, name, namespace=None):
+    """Decorator to manage running a node and shutting it down gracefully.
+
+    Note that this will wrap itself up cleanly and launch all nodes with a
+    single launcher, instead of multiples.
+    """
+    if not namespace:
+        namespace = '/'+package
+    def launcher(func):
+        @functools.wraps(func)
+        def new_test(self):
+            node = roslaunch.core.Node(package, name, namespace=namespace)
+            is_master = False
+            global _launcher
+            if _launcher is None:
+                launch = roslaunch.scriptapi.ROSLaunch()
+                launch.start()
+                _launcher = launch
+                is_master = True
+            else:
+                launch = _launcher
+
+            process = launch.launch(node)
+            # Beware this is a bit of a hack, and will currently not work if we
+            # want to run more than 1 node with the same name.
+            while not any(nn.split('/')[-1].startswith(name.replace('.', '_')) 
```

Comment:
so when you say "run more than one node with the same name", to me it sounds like you can't run multiple tests that test the same node at the same time. While this is probably okay, in the future we may want to have a way of namespacing tests themselves, so they don't interfere with eachother. You can even use the test name as the node namespace.